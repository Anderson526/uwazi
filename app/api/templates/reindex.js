import templates from './templates';

const SHOULD_NOT_TRIGGER_REINDEX = [
  'name',
  'color',
  'properties.filter',
  'properties.defaultfilter',
  'properties.noLabel',
  'properties.showInCard',
  'properties.required',
  'properties.style',
  'properties.fullWidth',
  'properties.nestedProperties',
  'properties.localID',
  'commonProperties.filter',
  'commonProperties.defaultfilter',
  'commonProperties.noLabel',
  'commonProperties.showInCard',
  'commonProperties.required',
  'commonProperties.prioritySorting',
];

function compareTemplateProperties(updatedProperties, originalProperties) {
  const changedProps = [];
  const hasNewProperty = updatedProperties.find(prop => !prop.hasOwnProperty('_id'));

  if (hasNewProperty) {
    changedProps.push('PROPERTY_ADDED');
  }

  originalProperties.forEach(originalProperty => {
    const updatedProperty = updatedProperties.find(prop => prop.name === originalProperty.name);

    if (!updatedProperty) {
      changedProps.push('PROPERTY_DELETED');
      return;
    }

    Object.keys(updatedProperty).forEach(originalKey => {
      if (updatedProperty[originalKey] !== originalProperty[originalKey] && originalKey !== '_id') {
        changedProps.push(originalKey);
      }
    });
  });

  return changedProps;
}
function checkIfFilterConditionsMet(changedProperties) {
  let shouldReindex = false;
  changedProperties.forEach(property => {
    if (!SHOULD_NOT_TRIGGER_REINDEX.includes(property)) {
      shouldReindex = true;
    }
  });
  return shouldReindex;
}

export async function checkIfReindex(updatedTemplate) {
  if (updatedTemplate._id) {
    const changedProperties = [];
    const [originalTemplate] = await templates.get({ _id: updatedTemplate._id });
    const updatedKeys = Object.keys(updatedTemplate);

    // Compare shallow object properties
    updatedKeys.forEach(key => {
      if (Array.isArray(updatedTemplate[key])) {
        // Is a list of properties
        const props = compareTemplateProperties(updatedTemplate[key], originalTemplate[key]);
        props.forEach(propsKey => {
          changedProperties.push(`${key}.${propsKey}`);
        });
        return;
      }
      if (updatedTemplate[key] !== originalTemplate[key] && key !== '_id') {
        changedProperties.push(key);
      }
    });
    return checkIfFilterConditionsMet(changedProperties);
  }
  return false;
}
