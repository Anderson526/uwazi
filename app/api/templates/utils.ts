import uuid from 'node-uuid';
import { ObjectID } from 'mongodb';
import { differenceBy } from 'lodash';

import settings from 'api/settings/settings';
import { files } from 'api/files';
import propertiesHelper from 'shared/comonProperties';
import { safeName as sharedSafeName } from 'shared/propertyNames';
import { ensure } from 'shared/tsUtils';
import { PropertySchema } from 'shared/types/commonTypes';
import { TemplateSchema } from 'shared/types/templateType';
import { ThesaurusValueSchema } from 'shared/types/thesaurusType';
import model from './templatesModel';

export const safeName = sharedSafeName;

const getInheritedProps = async (templates: TemplateSchema[]) => {
  const properties: PropertySchema[] = propertiesHelper
    .allUniqueProperties(templates)
    .filter((p: PropertySchema) => p.inherit?.property);

  return (
    await model.db.aggregate([
      {
        $match: {
          'properties._id': {
            $in: properties.map(p => new ObjectID(p.inherit?.property)),
          },
        },
      },
      {
        $project: {
          properties: {
            $filter: {
              input: '$properties',
              as: 'property',
              cond: {
                $or: properties.map(p => ({
                  $eq: ['$$property._id', new ObjectID(p.inherit?.property)],
                })),
              },
            },
          },
          _id: 0,
        },
      },
      { $unwind: '$properties' },
      { $replaceRoot: { newRoot: '$properties' } },
    ])
  ).reduce((indexed, prop) => ({ ...indexed, [prop._id.toString()]: prop }), {});
};

export const denormalizeInheritedProperties = async (template: TemplateSchema) => {
  const inheritedProperties: { [k: string]: PropertySchema } = await getInheritedProps([template]);

  return template.properties?.map(prop => {
    if (!prop.inherit?.property) {
      delete prop.inherit;
      return prop;
    }

    const { type } = inheritedProperties[prop.inherit.property];
    // eslint-disable-next-line no-param-reassign
    prop.inherit.type = type;
    return prop;
  });
};

const generateName = (property: PropertySchema, newNameGeneration: boolean) => {
  const name = property.label ? safeName(property.label, newNameGeneration) : property.name;
  return property.type === 'geolocation' || property.type === 'nested'
    ? `${name}_${property.type}`
    : name;
};

const generateNames = (properties: PropertySchema[], newNameGeneration: boolean) =>
  properties.map(property => ({
    ...property,
    name: generateName(property, newNameGeneration),
  }));

export function generateIds(properties: PropertySchema[] = []) {
  return properties.map(property => ({
    ...property,
    id: property.id || uuid.v4(),
  }));
}

export const generateNamesAndIds = async (properties: PropertySchema[] = []) => {
  const { newNameGeneration = false } = await settings.get();
  return generateIds(generateNames(properties, newNameGeneration));
};

export interface PropertyOrThesaurusSchema
  extends Partial<PropertySchema>,
    Partial<ThesaurusValueSchema> {}

const flattenProperties = (properties: PropertyOrThesaurusSchema[]) =>
  properties.reduce<PropertyOrThesaurusSchema[]>((flatProps, p) => {
    if (p.values) {
      return [...flatProps, ...p.values, p];
    }

    return [...flatProps, p];
  }, []);

export function getUpdatedNames(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[],
  prop: 'name' | 'label' = 'name',
  outKey: 'name' | 'label' = prop
) {
  const propertiesWithNewName: { [k: string]: string | undefined } = {};
  flattenProperties(oldProperties).forEach(property => {
    const newProperty = flattenProperties(newProperties).find(p => p.id === property.id);
    if (newProperty && newProperty[prop] !== property[prop]) {
      const key = property[outKey];
      if (key) {
        propertiesWithNewName[key] = newProperty[prop];
      }
    }
  });

  return propertiesWithNewName;
}

const notIncludedIn = (propertyCollection: PropertyOrThesaurusSchema[]) => (
  property: PropertyOrThesaurusSchema
) => !propertyCollection.find(p => p.id === property.id);

export function getDeletedProperties(
  oldProperties: PropertyOrThesaurusSchema[] = [],
  newProperties: PropertyOrThesaurusSchema[],
  prop: 'name' | 'label' = 'name'
) {
  return flattenProperties(oldProperties)
    .filter(notIncludedIn(flattenProperties(newProperties)))
    .map(property => property[prop]);
}

export function getRenamedTitle(
  oldCommonProperties: PropertySchema[],
  newCommonProperties: PropertySchema[]
) {
  const oldTitle = ensure<PropertySchema>(oldCommonProperties.find(p => p.name === 'title'));
  const newTitle = ensure<PropertySchema>(newCommonProperties.find(p => p.name === 'title'));
  return oldTitle.label !== newTitle.label ? [oldTitle.label] : [];
}

export const removeExtractedMetadata = async (
  oldProperties: PropertySchema[] = [],
  newProperties: PropertySchema[] = []
) => {
  const hasRemovedProperties = oldProperties.length > newProperties.length;
  let removedProperties: PropertySchema[] = [];

  if (hasRemovedProperties) {
    const difference = differenceBy(oldProperties, newProperties, 'name');
    removedProperties = difference.filter(property =>
      ['text', 'markdown', 'numeric', 'date'].includes(property.type)
    );
  }
  if (removedProperties.length > 0) {
    return removedProperties.reduce(async (previousPromise: Promise<void>, property) => {
      await previousPromise;
      const affectedFiles = await files.get({
        'extractedMetadata.propertyID': property._id?.toString(),
      });

      await affectedFiles.reduce(async (prevPromise: Promise<void>, file) => {
        await prevPromise;
        await files.save({
          ...file,
          extractedMetadata: file.extractedMetadata?.filter(
            data => data.propertyID !== property._id?.toString()
          ),
        });
      }, Promise.resolve());
    }, Promise.resolve());
  }

  return null;
};
