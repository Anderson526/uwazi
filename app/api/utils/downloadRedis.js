/* eslint-disable no-console */
//eslint-disable-next-line node/no-restricted-import
import fs from 'fs';
import { execSync } from 'child_process';
import path from 'path';

export const downloadRedis = () => {
  const redis_version = (execSync('cat .redis_version').toString()).replace('\n', '');
  const pathToBin = path.join(__dirname, `../../../redis-bin/redis-stable/src/redis-server`);
  console.log(pathToBin);
  if (fs.existsSync(pathToBin)) {
    return;
  }
  console.log('Downloading redis...');

  execSync('mkdir -p redis-bin', { stdio: 'ignore' });
  execSync(`cd redis-bin && curl -O http://download.redis.io/releases/redis-${redis_version}.tar.gz` , { stdio: 'ignore' });
  execSync(`cd redis-bin && tar xzvf redis-${redis_version}.tar.gz`, { stdio: 'ignore' });
  execSync(`cd redis-bin && mv redis-${redis_version} redis-stable`, { stdio: 'ignore' });

  console.log('Downloading redis... Done');
  console.log('Installing redis...');
  execSync(
    `cd redis-bin &&
       cd redis-stable &&
       make distclean &&
       make`,
    { stdio: 'ignore' }
  );
  console.log('Installing redis... Done');
};
