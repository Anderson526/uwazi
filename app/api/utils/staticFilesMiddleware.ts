import { NextFunction, Request, Response } from 'express';
import { pathFunction, fileExists } from 'api/files';
import { createError } from '.';

const staticFilesMiddleware = (pathFunctions: pathFunction[]) => async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  const pathToUse = await pathFunctions.reduce<Promise<pathFunction>>(async (current, path) => {
    if (await fileExists(path(req.params.fileName))) {
      return path;
    }
    return current;
  }, Promise.resolve(pathFunctions[0]));
  try {
    if (!req.params.fileName || !(await fileExists(pathToUse(req.params.fileName)))) {
      throw createError('file not found', 404);
    }
    res.sendFile(pathToUse(req.params.fileName));
  } catch (e) {
    next();
  }
};

export { staticFilesMiddleware };
