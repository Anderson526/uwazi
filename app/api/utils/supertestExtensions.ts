import { Response } from 'supertest';
// @ts-ignore
import Test from 'supertest/lib/test';

function extractStatusDebugInfo(res: Response): string {
  switch (res.status) {
    case 400:
      return JSON.stringify(JSON.parse(res.text), null, 2);
    default:
      return '';
  }
}

export function extendSupertest() {
  const { _assertStatus } = Test.prototype;

  Test.prototype._assertStatus = function extendedAssertStatus(status: number, res: Response) {
    const err: Error = _assertStatus(status, res);
    if (err) {
      err.message += `\n ${extractStatusDebugInfo(res)}`;
      delete err.stack;
    }
    return err;
  };
}
