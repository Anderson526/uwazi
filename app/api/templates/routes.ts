/* eslint-disable max-statements */
import { Application } from 'express';
import settings from 'api/settings';
import { checkMapping, reindexAll } from 'api/search/entitiesIndex';
import { search } from 'api/search';
import { validation } from '../utils';
import needsAuthorization from '../auth/authMiddleware';
import templates from './templates';
import { generateNamesAndIds } from './utils';
import { checkIfReindex } from './reindex';

export default (app: Application) => {
  app.post('/api/templates', needsAuthorization(), async (req, res, next) => {
    try {
      const { reindex: fullReindex } = req.body;
      delete req.body.reindex;
      const templateProperties = await generateNamesAndIds(req.body.properties);
      const template = { ...req.body, properties: templateProperties };
      const { valid, error } = await checkMapping(template);

      if (!valid && !fullReindex) {
        return res.json({ error: `Reindex requiered: ${error}` });
      }

      const reindex = fullReindex ? false : await checkIfReindex(req.body);
      const response = await templates.save(req.body, req.language, reindex);

      req.sockets.emitToCurrentTenant('templateChange', response);

      const updatedSettings = await settings.updateFilterName(
        response._id.toString(),
        response.name
      );

      if (updatedSettings) {
        req.sockets.emitToCurrentTenant('updateSettings', updatedSettings);
      }

      if (fullReindex) {
        const allTemplates = await templates.get();
        await reindexAll(allTemplates, search);
      }

      return res.json(response);
    } catch (error) {
      next(error);
    }
  });

  // app.post('/api/templates/check_mapping', needsAuthorization(), async (req, res, next) => {
  //   const template = req.body;
  //   template.properties = await generateNamesAndIds(template.properties);
  //   checkMapping(template)
  //     .then(response => res.json(response))
  //     .catch(next);
  // });

  app.post(
    '/api/templates/setasdefault',
    needsAuthorization(),
    validation.validateRequest({
      properties: {
        body: {
          properties: {
            _id: { type: 'string' },
          },
        },
      },
    }),
    async (req, res, next) => {
      try {
        const [newDefault, oldDefault] = await templates.setAsDefault(req.body._id.toString());
        req.sockets.emitToCurrentTenant('templateChange', newDefault);
        if (oldDefault) {
          req.sockets.emitToCurrentTenant('templateChange', oldDefault);
        }
        res.json(newDefault);
      } catch (err) {
        next(err);
      }
    }
  );

  app.get('/api/templates', (_req, res, next) => {
    templates
      .get()
      .then(response => res.json({ rows: response }))
      .catch(next);
  });

  app.delete(
    '/api/templates',
    needsAuthorization(),
    validation.validateRequest({
      properties: {
        query: {
          properties: {
            _id: { type: 'string' },
          },
        },
      },
    }),
    (req, res, next) => {
      const template = { _id: req.query._id, name: req.query.name };
      templates
        .delete(template)
        .then(async () => settings.removeTemplateFromFilters(template._id))
        .then(newSettings => {
          res.json(template);
          req.sockets.emitToCurrentTenant('updateSettings', newSettings);
          req.sockets.emitToCurrentTenant('templateDelete', template);
        })
        .catch(next);
    }
  );

  app.get(
    '/api/templates/count_by_thesauri',
    validation.validateRequest({
      properties: {
        query: {
          properties: {
            _id: { type: 'string' },
          },
        },
      },
    }),
    (req, res, next) => {
      templates
        .countByThesauri(req.query._id)
        .then(response => res.json(response))
        .catch(next);
    }
  );
};
