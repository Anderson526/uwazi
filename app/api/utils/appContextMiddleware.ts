import { Request, Response, NextFunction } from 'express';
import { appContext } from 'api/utils/AppContext';

const appContextMiddleware = (_req: Request, _res: Response, next: NextFunction) => {
  appContext
    .run(async () => {
      appContext.set('requestId', Math.floor(Math.random() * 10000));
      next();
    })
    .catch(next);
};

export { appContextMiddleware };
