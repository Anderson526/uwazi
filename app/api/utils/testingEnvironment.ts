import { UserInContextMockFactory } from 'api/utils/testingUserInContext';
import { appContext } from 'api/utils/AppContext';
import testingDB, { DBFixture } from 'api/utils/testing_db';
import { UserSchema } from 'shared/types/userType';
import { elasticTesting } from './elastic_testing';
import { testingTenants } from './testingTenants';

const appContextTestValues: { [k: string]: any } = {
  requestId: '1234',
};

interface TestingEnvironment {
  promises: Promise<any>[];
  appContextSpy: jest.SpyInstance | null;
  setUp: (fixtures: DBFixture, elasticIndex?: string) => Promise<void>;

  connect(): TestingEnvironment;
  disconnect(): void;

  withFixtures(fixtures: DBFixture): void;

  withElastic(elasticIndex: string): void;

  withContext(): any;

  withPermissions(): void;

  run(): void;
}

const testingEnvironment: TestingEnvironment = {
  promises: [],
  appContextSpy: null,
  async setUp(fixtures?: DBFixture, elasticIndex?: string) {
    if (fixtures) {
      this.connect();
      this.withFixtures(fixtures);
    }
    if (elasticIndex) {
      this.withElastic(elasticIndex);
    }
    this.withPermissions();
    this.withContext();
    await this.run();
  },

  async run() {
    await Promise.all(this.promises);
  },
  connect() {
    this.promises.push(testingDB.connect());
    return this;
  },
  async disconnect() {
    await testingDB.disconnect();
  },
  withFixtures(fixtures?: DBFixture) {
    if (fixtures) {
      this.promises.push(testingDB.clearAllAndLoad(fixtures));
    }
    return this;
  },
  withElastic(elasticIndex: string) {
    testingTenants.changeCurrentTenant({ indexName: elasticIndex });
    this.promises.push(elasticTesting.reindex());
    return this;
  },
  withPermissions(user?: UserSchema) {
    const userInContextMockFactory = new UserInContextMockFactory();
    if (!user) {
      userInContextMockFactory.mockEditorUser();
    } else {
      userInContextMockFactory.mock(user);
    }
    return this;
  },
  withContext(contextValues: { [k: string]: any } = appContextTestValues) {
    this.appContextSpy = jest
      .spyOn(appContext, 'get')
      .mockImplementation((key: string) => contextValues[key]);
    return this;
  },
};

export { testingEnvironment };
