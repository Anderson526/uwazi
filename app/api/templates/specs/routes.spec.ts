import { Application, NextFunction } from 'express';
import request from 'supertest';
import { setUpApp } from 'api/utils/testingRoutes';
import { testingEnvironment } from 'api/utils/testingEnvironment';
import { getFixturesFactory } from 'api/utils/fixturesFactory';
import translations from 'api/i18n';
import { UserRole } from 'shared/types/userSchema';
import templateRoutes from '../routes';
import templates from '../templates';

jest.mock(
  '../../auth/authMiddleware.ts',
  () => () => (_req: Request, _res: Response, next: NextFunction) => {
    next();
  }
);

jest.mock(
  '../../utils/languageMiddleware.ts',
  () => (_req: Request, _res: Response, next: NextFunction) => {
    next();
  }
);

const templateCommonProperties = [
  {
    _id: '6193bf8c86a5e87060962287',
    localID: 'commonTitle',
    label: 'Title',
    name: 'title',
    isCommonProperty: true,
    type: 'text',
    prioritySorting: false,
    generatedId: false,
  },
  {
    _id: '6193bf8c86a5e87060962288',
    localID: 'commonCreationDate',
    label: 'Date added',
    name: 'creationDate',
    isCommonProperty: true,
    type: 'date',
    prioritySorting: false,
  },
  {
    _id: '6193bf8c86a5e87060962289',
    localID: 'commonEditDate',
    label: 'Date modified',
    name: 'editDate',
    isCommonProperty: true,
    type: 'date',
    prioritySorting: false,
  },
];

const fixtureFactory = getFixturesFactory();
const fixtures = {
  templates: [
    {
      ...fixtureFactory.template('template1', []),
      commonProperties: templateCommonProperties,
    },
    fixtureFactory.template('template2', []),
    fixtureFactory.template('template3', []),
  ],
};

const emitToCurrentTenantSpy = jasmine.createSpy('emitToCurrentTenant');

describe('templates routes', () => {
  const app: Application = setUpApp(templateRoutes, (req, _res, next: NextFunction) => {
    (req as any).user = {
      role: UserRole.ADMIN,
      username: 'admin',
    };
    req.sockets = { emitToCurrentTenant: emitToCurrentTenantSpy };
    next();
  });

  const postToEnpoint = async (route: string, body: any) =>
    request(app)
      .post(route)
      .send(body)
      .expect(200);

  beforeEach(async () => {
    await testingEnvironment.setUp(fixtures, 'templates_index');
    spyOn(translations, 'updateContext').and.returnValue(Promise.resolve());
  });

  afterAll(async () => testingEnvironment.tearDown());

  describe('GET', () => {
    it('should return all templates by default', async () => {
      const { body } = await request(app)
        .get('/api/templates')
        .expect(200);

      expect(JSON.stringify(body.rows)).toBe(JSON.stringify(fixtures.templates));
    });
  });

  describe('DELETE', () => {
    it('should delete a template', async () => {
      await request(app)
        .delete(`/api/templates?_id=${fixtureFactory.template('template2', [])._id}`)
        .expect(200);
      const remainingTemplates = await templates.get();
      expect(remainingTemplates).toContainEqual(expect.objectContaining({ name: 'template1' }));
      expect(remainingTemplates).toContainEqual(expect.objectContaining({ name: 'template3' }));
      expect(remainingTemplates).not.toContainEqual(expect.objectContaining({ name: 'template2' }));
    });

    it('should validate that request has _id', async () => {
      const { body } = await request(app)
        .delete('/api/templates')
        .expect(400);
      expect(body.error).toBe('validation failed');
    });
  });

  describe('POST', () => {
    it('should create a template', async () => {
      const templateToSave = {
        name: 'template4',
        properties: [],
        commonProperties: templateCommonProperties,
      };

      await postToEnpoint('/api/templates', templateToSave);

      const savedTemplates = await templates.get();

      expect(savedTemplates).toContainEqual(expect.objectContaining({ name: 'template4' }));
    });

    it('should update a existing template', async () => {
      const [firstTemplate] = await templates.get();
      const templateToUpdate = {
        ...firstTemplate,
        properties: [{ label: 'Numeric', type: 'numeric', localID: 'z0x8wx8xy4' }],
        commonProperties: templateCommonProperties,
        __v: 0,
      };

      await postToEnpoint('/api/templates', templateToUpdate);

      const [updatedTemplate] = await templates.get({ _id: templateToUpdate._id });
      expect(updatedTemplate.properties).toContainEqual(
        expect.objectContaining({ label: 'Numeric', type: 'numeric', localID: 'z0x8wx8xy4' })
      );
    });

    it('should not emit settings update when settings not modified', async () => {
      const templateToSave = {
        name: 'new template',
        properties: [],
        commonProperties: templateCommonProperties,
      };

      await postToEnpoint('/api/templates', templateToSave);

      expect(emitToCurrentTenantSpy).not.toHaveBeenCalledWith('updateSettings');
    });
  });

  describe('/templates/count_by_thesauri', () => {
    it('should have a validation schema', () => {});
    it('should return the number of templates using a thesauri', () => {});
  });

  describe('/api/templates/setasdefault', () => {
    it('should have a validation schema', () => {});

    it('should call templates to set the new default', () => {});
  });

  describe('check mappings', () => {
    it('should check if a template is valid vs the current elasticsearch mapping', () => {});
  });
});
