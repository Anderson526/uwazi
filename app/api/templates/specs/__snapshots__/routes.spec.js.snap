// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates routes /api/templates/setasdefault should have a validation schema 1`] = `undefined`;

exports[`templates routes /templates/count_by_thesauri should have a validation schema 1`] = `undefined`;

exports[`templates routes DELETE should have a validation schema 1`] = `undefined`;
